File: src/extensibility/dotnet-extensibility/DataModel/ResourceBuilder.cs
using System.Diagnostics;
using System.Text.Json;

namespace Nexus.DataModel;

/// <summary>
/// A resource builder simplifies building a resource.
/// </summary>
[DebuggerDisplay("{Id,nq}")]
public record ResourceBuilder
{
    private readonly string _id;
    private Dictionary<string, JsonElement>? _properties;
    private List<Representation>? _representations;

    /// <summary>
    /// Initializes a new instance of the <see cref="ResourceBuilder"/>.
    /// </summary>
    /// <param name="id">The identifier of the resource to be built.</param>
    public ResourceBuilder(string id)
    {
        _id = id;
    }

    /// <summary>
    /// Adds a property.
    /// </summary>
    /// <param name="key">The key of the property.</param>
    /// <param name="value">The value of the property.</param>
    /// <returns>The resource builder.</returns>
    public ResourceBuilder WithProperty(string key, object value)
    {
        _properties ??= new();

        _properties[key] = JsonSerializer.SerializeToElement(value);

        return this;
    }

    /// <summary>
    /// Adds a <see cref="Representation"/>.
    /// </summary>
    /// <param name="representation">The <see cref="Representation"/>.</param>
    /// <returns>The resource builder.</returns>
    public ResourceBuilder AddRepresentation(Representation representation)
    {
        _representations ??= new List<Representation>();

        _representations.Add(representation);

        return this;
    }

    /// <summary>
    /// Adds a list of <see cref="Representation"/>.
    /// </summary>
    /// <param name="representations">The list of <see cref="Representation"/>.</param>
    /// <returns>The resource builder.</returns>
    public ResourceBuilder AddRepresentations(params Representation[] representations)
    {
        return AddRepresentations((IEnumerable<Representation>)representations);
    }

    /// <summary>
    /// Adds a list of <see cref="Representation"/>.
    /// </summary>
    /// <param name="representations">The list of <see cref="Representation"/>.</param>
    /// <returns>The resource builder.</returns>
    public ResourceBuilder AddRepresentations(IEnumerable<Representation> representations)
    {
        _representations ??= new List<Representation>();

        _representations.AddRange(representations);

        return this;
    }

    /// <summary>
    /// Builds the <see cref="Resource"/>.
    /// </summary>
    /// <returns>The <see cref="Resource"/>.</returns>
    public Resource Build()
    {
        return new Resource(_id, _properties, _representations);
    }
}

File: src/extensibility/rust-extensibility/src/data_model/resource_builder.rs
use std::collections::HashMap;

use super::{
    Representation, Resource,
    resource::{Representations, RepresentationsError, ResourceId},
};

const DESCRIPTION: &str = "description";
const WARNING: &str = "warning";
const UNIT: &str = "unit";
const GROUPS: &str = "groups";

/// A resource builder simplifies building a resource.
pub struct ResourceBuilder {
    id: ResourceId,
    properties: Option<HashMap<String, String>>,
    representations: Option<Vec<Representation>>,
}

impl ResourceBuilder {
    /// Initializes a new instance of the ResourceBuilder
    pub fn new(id: ResourceId) -> Self {
        ResourceBuilder {
            id,
            properties: None,
            representations: None,
        }
    }

    /// Adds a property.
    pub fn with_property(&mut self, key: String, value: String) -> &mut Self {
        self.properties
            .get_or_insert_with(HashMap::new)
            .insert(key, value);

        self
    }

    /// Adds a unit.
    pub fn with_unit(&mut self, unit: String) -> &mut Self {
        self.with_property(UNIT.to_string(), unit);
        self
    }

    /// Adds a description.
    pub fn with_description(&mut self, description: String) -> &mut Self {
        self.with_property(DESCRIPTION.to_string(), description);
        self
    }

    /// Adds a warning.
    pub fn with_warning(&mut self, warning: String) -> &mut Self {
        self.with_property(WARNING.to_string(), warning);
        self
    }

    /// Adds groups.
    pub fn with_groups(&mut self, groups: Vec<String>) -> &mut Self {
        let groups_value = groups.join(",");
        self.with_property(GROUPS.to_string(), groups_value);
        self
    }

    /// Adds a representation.
    pub fn add_representation(&mut self, representation: Representation) -> &mut Self {
        if self.representations.is_none() {
            self.representations = Some(vec![]);
        }

        self.representations.as_mut().unwrap().push(representation);
        self
    }

    /// Adds a list of representations.
    pub fn add_representations(&mut self, representations: Vec<Representation>) -> &mut Self {
        self.representations
            .get_or_insert_with(Vec::new)
            .extend(representations);

        self
    }

    /// Builds the resource.
    pub fn build(self) -> Result<Resource, RepresentationsError> {
        let representations = self
            .representations
            .map(Representations::try_new)
            .transpose()?;

        Ok(Resource {
            id: self.id,
            properties: self.properties,
            representations,
        })
    }
}
